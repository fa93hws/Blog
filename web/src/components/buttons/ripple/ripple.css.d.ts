// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'blue': string;
  'colorBlue': string;
  'gridBaseline': string;
  'monoWhite': string;
  'ripple': string;
  'rippleEffect': string;
  'white': string;
}
export const cssExports: CssExports;
export default cssExports;
