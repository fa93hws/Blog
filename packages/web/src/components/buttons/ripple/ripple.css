@value gridBaseline from '~components/metrics.css';
@value colorBlue from '~components/colors.css';
@value monoWhite from '~components/colors.css';

.ripple {
  position: relative;
  border-radius: 50%;
  width: calc(gridBaseline * 0.5);
  height: calc(gridBaseline * 0.5);
  opacity: 0;
  animation: rippleEffect 500ms 1;
}

.white {
  background-color: monoWhite;
}

.blue {
  background-color: colorBlue;
}

@keyframes rippleEffect {
  0% {
    transform: scale(1);
    opacity: 0.4;
  }
  100% {
    transform: scale(100);
    opacity: 0;
  }
}
