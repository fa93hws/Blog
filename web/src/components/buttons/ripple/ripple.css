@value gridBaseline from '~components/metrics.css';
@value monoWhite from '~components/colors.css';

.ripple {
  position: absolute;
  background: monoWhite;
  border-radius: 50%;
  width: calc(gridBaseline * 0.5);
  height: calc(gridBaseline * 0.5);
  pointer-events: none;
}

.ripple {
  opacity: 0.4;
  transform: scale(10);
}

.ripple.useAnimation {
  opacity: 0;
  animation: rippleEffect 1500ms 1;
}

@keyframes rippleEffect {
  0% {
    transform: scale(1);
    opacity: 0.4;
  }
  100% {
    transform: scale(100);
    opacity: 0;
  }
}
