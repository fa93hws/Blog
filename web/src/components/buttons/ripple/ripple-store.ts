import { observable, action } from 'mobx';
import { RipplePosition } from './ripple';

export class RippleStore {
  @observable.ref
  ripplePosition: RipplePosition | undefined;

  @action
  setRipple(position: RipplePosition) {
    this.ripplePosition = position;
  }
}
