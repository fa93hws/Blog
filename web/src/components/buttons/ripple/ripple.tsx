import * as React from 'react';
import { observer } from 'mobx-react';
import * as classnames from 'classnames';
import styles from './ripple.css';
import { RippleStore } from './ripple-store';

export type RipplePosition = {
  top: number;
  left: number;
};
export type RippleColor = 'white' | 'blue';

export namespace Internal {
  export const Ripple = React.memo(
    ({ position, color }: { position: RipplePosition; color: RippleColor }) => (
      <div
        key={Math.random()}
        className={classnames(styles.ripple, styles[color])}
        style={{ top: position.top, left: position.left }}
      />
    ),
  );
}

export type RipplableProps = {
  onMouseDown(e: React.MouseEvent<HTMLElement, MouseEvent>): void;
  RippleSlot: React.ComponentType<{ color: RippleColor }>;
};

const defaultStoreFactory = () => new RippleStore();
export function withRipple<T extends RipplableProps>(
  Container: React.ComponentType<T>,
  storeFactory: () => RippleStore = defaultStoreFactory,
) {
  return React.memo((originalProps: Omit<T, keyof RipplableProps>) => {
    const store = React.useMemo(storeFactory, []);
    const onMouseDown = React.useCallback(
      (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        const { left, top } = e.currentTarget.getBoundingClientRect();
        const offsetY = e.clientY - top;
        const offsetX = e.clientX - left;
        store.setRipple({ top: offsetY, left: offsetX });
      },
      [],
    );

    const RippleImpl = React.useMemo(
      () =>
        observer(({ color }: { color: RippleColor }) =>
          store.ripplePosition != null ? (
            <Internal.Ripple position={store.ripplePosition} color={color} />
          ) : null,
        ),
      [],
    );

    return (
      <Container
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        {...(originalProps as any)}
        onMouseDown={onMouseDown}
        RippleSlot={RippleImpl}
      />
    );
  });
}
