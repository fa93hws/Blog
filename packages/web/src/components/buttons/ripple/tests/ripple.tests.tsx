import * as React from 'react';
import { mount } from 'enzyme';
import { Internal, withRipple, RipplableProps, RippleColor } from '../ripple';
import { RippleStore } from '../ripple-store';

describe('Ripple', () => {
  function createContainer(color: RippleColor) {
    return (props: RipplableProps) => (
      <button onMouseDown={props.onMouseDown}>
        button text
        <props.RippleSlot color={color} />
      </button>
    );
  }

  it('renders the white ripple', () => {
    const { Ripple } = Internal;
    expect(
      <Ripple position={{ top: 2, left: 4 }} color="white" />,
    ).toMatchRenderedSnapshot();
  });

  it('renders the blue ripple', () => {
    const { Ripple } = Internal;
    expect(
      <Ripple position={{ top: 2, left: 4 }} color="blue" />,
    ).toMatchRenderedSnapshot();
  });

  it('renders the container without ripple', () => {
    const Container = createContainer('white');
    const RippleComponent = withRipple(Container);
    expect(<RippleComponent />).toMatchRenderedSnapshot();
  });

  it('renders the container with white ripple', () => {
    const Container = createContainer('white');
    const storeFactory = () => {
      const store = new RippleStore();
      store.setRipple({ top: 2, left: 4 });
      return store;
    };
    const RippleComponent = withRipple(Container, storeFactory);
    expect(<RippleComponent />).toMatchRenderedSnapshot();
  });

  it('triggers ripple when mouse down', () => {
    const Container = createContainer('white');
    const store = new RippleStore();
    const storeFactory = () => store;
    const RippleComponent = withRipple(Container, storeFactory);
    const wrapper = mount(<RippleComponent />);
    wrapper.find('button').simulate('mouseDown', { clientX: 10, clientY: 20 });
    expect(store.ripplePosition).toEqual({
      left: 10,
      top: 20,
    });
  });
});
