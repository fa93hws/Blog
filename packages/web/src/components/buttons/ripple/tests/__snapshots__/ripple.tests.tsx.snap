// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple renders the blue ripple 1`] = `
<div
  class="ripple blue"
  style="top:2px;left:4px"
/>
`;

exports[`Ripple renders the container with white ripple 1`] = `
<button>
  button text
  <div
    class="ripple white"
    style="top:2px;left:4px"
  />
</button>
`;

exports[`Ripple renders the container without ripple 1`] = `
<button>
  button text
</button>
`;

exports[`Ripple renders the white ripple 1`] = `
<div
  class="ripple white"
  style="top:2px;left:4px"
/>
`;
